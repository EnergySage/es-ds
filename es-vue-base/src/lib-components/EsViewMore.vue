<template>
    <div>
        <!-- eslint-disable-next-line vue/no-v-html -->
        <span v-html="bodyContent" />
        <EsButton
            v-if="isTruncated"
            variant="link"
            class="p-0"
            @click="click">
            {{ buttonText }}
        </EsButton>
    </div>
</template>

<script lang="js">

import EsButton from '@/src/lib-components/EsButton.vue';
import truncate from 'truncate-html';

export default {
    name: 'EsViewMore',
    components: {
        EsButton,
    },
    props: {
        /**
         * Content of the button
         */
        content: {
            type: String,
            required: true,
        },
        /**
         * Maximum character length of text before it is truncated
         */
        length: {
            type: Number,
            required: true,
        },
        /**
         * Text for button to show more text
         */
        clamp: {
            type: String,
            required: false,
            default: 'Show More',
        },
        /**
         * Text for button to hide text
         */
        less: {
            type: String,
            required: false,
            default: 'Show Less',
        },
        /**
         * Choose whether to break on word or character
         */
        break: {
            type: String,
            required: false,
            default: 'character',
            validator: (val) => ['wordCount', 'wordFit', 'character'].includes(val),
        },
    },
    data() {
        return {
            isClamped: true,
        };
    },
    computed: {
        bodyContent() {
            if (this.isClamped) {
                return this.truncatedContent;
            }
            return this.content;
        },
        truncatedContent() {
            switch (this.break) {
            case 'character':
                return truncate(this.content, this.length);
            case 'wordCount':
                return truncate(this.content, this.length, { byWords: true });
            case 'wordFit':
                return truncate(this.content, this.wordBreakIndex, { byWords: true });
            default:
                return this.content;
            }
        },
        isTruncated() {
            return this.truncatedContent !== this.content;
        },
        buttonText() {
            return this.isClamped ? this.clamp : this.less;
        },
        // Calculates the amount of words that can fit within this.length
        wordBreakIndex() {
            let wordCount = 0;
            let totalWordLength = 0;
            const words = this.content.split(' ');
            // Iterate through word list until we hit a value that would overflow
            // our desired character count
            words.every((word) => {
                if (word.length + totalWordLength <= this.length) {
                    totalWordLength += word.length;
                    wordCount += 1;

                    // Accounts for the space between the previous word and new word
                    if (wordCount > 1) {
                        totalWordLength += 1;
                    }
                    return true;
                }
                return false;
            });

            // Must return at least 1 word for truncation to work properly
            return Math.max(1, wordCount);
        },
    },
    methods: {
        click() {
            if (this.$listeners && this.$listeners.click) {
                /**
                     * click event.
                     *
                     * @event click
                */
                this.$emit('click');
            } else {
                this.isClamped = !this.isClamped;
            }
        },
    },
};
</script>
