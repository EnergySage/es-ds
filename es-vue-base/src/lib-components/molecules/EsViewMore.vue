<template>
    <div>
        <!-- eslint-disable-next-line vue/no-v-html -->
        <span v-html="bodyContent" />
        <EsButton
            v-if="isTruncated"
            variant="link"
            class="p-0"
            @click="click">
            {{ buttonText }}
        </EsButton>
    </div>
</template>

<script lang="js">

import EsButton from '@/src/lib-components/molecules/EsButton.vue';
import truncate from 'html-truncate';

export default {
    name: 'EsViewMore',
    components: {
        EsButton,
    },
    props: {
        /**
         * Content of the button
         */
        content: {
            type: String,
            required: true,
        },
        /**
         * Maximum character length of text before it is truncated
         */
        length: {
            type: Number,
            required: true,
        },
        /**
         * Text for button to show more text
         */
        clamp: {
            type: String,
            required: false,
            default: 'Show More',
        },
        /**
         * Text for button to hide text
         */
        less: {
            type: String,
            required: false,
            default: 'Show Less',
        },
    },
    data() {
        return {
            isClamped: true,
        };
    },
    computed: {
        bodyContent() {
            if (this.isClamped) {
                return this.truncatedContent;
            }
            return this.content;
        },
        truncatedContent() {
            return truncate(this.content, this.length);
        },
        isTruncated() {
            return this.truncatedContent !== this.content;
        },
        buttonText() {
            return this.isClamped ? this.clamp : this.less;
        },
    },
    methods: {
        click() {
            if (this.$listeners && this.$listeners.click) {
                /**
                     * click event.
                     *
                     * @event click
                */
                this.$emit('click');
            } else {
                this.isClamped = !this.isClamped;
            }
        },
    },
};
</script>
