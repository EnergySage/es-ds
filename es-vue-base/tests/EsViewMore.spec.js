import { mount } from '@vue/test-utils';
import EsViewMore from '@/src/lib-components/EsViewMore.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsViewMore', () => {
    const defaultText = 'Hello world!';
    const htmlText = '<h1>Hello world!</h1>';

    // Basic test; should exist for most components
    test('<EsViewMore />', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                length: 20,
                content: 'Lorem ipsum dolor amet.',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    // =========== Basic Tests ===========
    test('wordFit truncation breaks at correct spot', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: defaultText,
                length: 6,
                break: 'wordFit',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.text()).toMatch('Hello...');
    });

    test('character truncation breaks at correct spot', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: defaultText,
                length: 8,
                break: 'character',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.text()).toMatch('Hello wo...');
    });

    test('text shorter than length is not truncated', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: defaultText,
                length: 20,
                break: 'character',
            },
        });

        // If there is no truncation, we don't show clamp button
        expect(wrapper.find('button').exists()).toBe(false);

        expect(wrapper.text()).toMatch(defaultText);
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('clicking "Show More" button reveals more text', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: defaultText,
                length: 6,
                break: 'wordFit',
            },
        });
        const showMoreButton = wrapper.findComponent('button');

        // Confirm starts truncated
        expect(wrapper.text()).toMatch('Hello...');

        await showMoreButton.trigger('click');

        // After click, all text is shown
        expect(wrapper.text()).toMatch(defaultText);
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('clicking "Show Less" button hides text to the break point', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: defaultText,
                length: 6,
                break: 'wordFit',
            },
        });
        const showMoreButton = wrapper.findComponent('button');

        // Starts truncated
        expect(wrapper.text()).toMatch('Hello...');

        // Click to show all text
        await showMoreButton.trigger('click');

        // After click, all text is shown
        expect(wrapper.text()).toMatch(defaultText);

        // Click again to hide text
        await showMoreButton.trigger('click');

        // Confirm text hidden again
        expect(wrapper.text()).toMatch('Hello...');
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('<EsViewMore clamp="testMore"', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: defaultText,
                length: 6,
                clamp: 'testMore',
            },
        });

        const showMoreButton = wrapper.findComponent('button');

        expect(showMoreButton.text()).toMatch('testMore');
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('<EsViewMore less="testLess', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: defaultText,
                length: 6,
                less: 'testLess',
            },
        });

        const showMoreButton = wrapper.findComponent('button');

        await showMoreButton.trigger('click');

        expect(showMoreButton.text()).toMatch('testLess');
        expect(wrapper.html()).toMatchSnapshot();
    });

    // =========== HTML truncation tests ===========
    test('character truncation with HTML tags only truncates text content', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: htmlText,
                length: 1,
                break: 'character',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.text()).toMatch('H...');
        expect(wrapper.html()).toMatch('<h1>H...</h1>');
    });

    test('wordFit truncation with HTML tags only truncates text content', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: htmlText,
                length: 1,
                break: 'wordFit',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.text()).toMatch('Hello...');
        expect(wrapper.html()).toMatch('<h1>Hello...</h1>');
    });

    test('wordCount truncation with HTML tags only truncates text content', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: htmlText,
                length: 1,
                break: 'wordCount',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.text()).toMatch('Hello...');
        expect(wrapper.html()).toMatch('<h1>Hello...</h1>');
    });

    // =========== wordFit break edge case tests ===========
    test('wordFit truncation works with one character', async () => {
        const oneCharacterText = 'H';
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: oneCharacterText,
                length: 10,
                break: 'wordFit',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.text()).toMatch('H');
    });

    test('wordFit truncation with only one word that is longer than length will render word', async () => {
        const oneWordText = 'Hello';
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: oneWordText,
                length: 1,
                break: 'wordFit',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.text()).toMatch('Hello');
    });

    // =========== $emit tests ===========
    test('"Show more" button emits on click', async () => {
        const click = jest.fn();
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            listeners: {
                click,
            },
            propsData: {
                content: defaultText,
                length: 10,
            },
        });

        const showMoreButton = wrapper.findComponent('button');

        showMoreButton.trigger('click');

        await wrapper.vm.$nextTick();

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.emitted('click')).toBeTruthy();
    });

    test('"Show more" does not emit if not listening for it', async () => {
        const wrapper = mount(EsViewMore, {
            ...jestVue,
            propsData: {
                content: defaultText,
                length: 10,
            },
        });

        const showMoreButton = wrapper.findComponent('button');

        showMoreButton.trigger('click');

        await wrapper.vm.$nextTick();

        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.emitted('click')).toBeFalsy();
    });
});
