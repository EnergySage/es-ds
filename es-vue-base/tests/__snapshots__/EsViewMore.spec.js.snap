// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsViewMore "Show more" button emits on click 1`] = `
"<div><span>Hello worl...</span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show More
  </button></div>"
`;

exports[`EsViewMore "Show more" does not emit if not listening for it 1`] = `
"<div><span>Hello world!</span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show Less
  </button></div>"
`;

exports[`EsViewMore <EsViewMore /> 1`] = `
"<div><span>Lorem ipsum dolor am...</span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show More
  </button></div>"
`;

exports[`EsViewMore <EsViewMore clamp="testMore" 1`] = `
"<div><span>Hello ...</span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    testMore
  </button></div>"
`;

exports[`EsViewMore <EsViewMore less="testLess 1`] = `
"<div><span>Hello world!</span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    testLess
  </button></div>"
`;

exports[`EsViewMore character truncation breaks at correct spot 1`] = `
"<div><span>Hello wo...</span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show More
  </button></div>"
`;

exports[`EsViewMore character truncation with HTML tags only truncates text content 1`] = `
"<div><span><h1>H...</h1></span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show More
  </button></div>"
`;

exports[`EsViewMore clicking "Show Less" button hides text to the break point 1`] = `
"<div><span>Hello...</span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show More
  </button></div>"
`;

exports[`EsViewMore clicking "Show More" button reveals more text 1`] = `
"<div><span>Hello world!</span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show Less
  </button></div>"
`;

exports[`EsViewMore text shorter than length is not truncated 1`] = `
"<div><span>Hello world!</span>
  <!---->
</div>"
`;

exports[`EsViewMore wordCount truncation with HTML tags only truncates text content 1`] = `
"<div><span><h1>Hello...</h1></span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show More
  </button></div>"
`;

exports[`EsViewMore wordFit truncation breaks at correct spot 1`] = `
"<div><span>Hello...</span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show More
  </button></div>"
`;

exports[`EsViewMore wordFit truncation with HTML tags only truncates text content 1`] = `
"<div><span><h1>Hello...</h1></span> <button type=\\"button\\" class=\\"btn p-0 btn-link btn-md\\">
    Show More
  </button></div>"
`;

exports[`EsViewMore wordFit truncation with only one word that is longer than length will render word 1`] = `
"<div><span>Hello</span>
  <!---->
</div>"
`;

exports[`EsViewMore wordFit truncation works with one character 1`] = `
"<div><span>H</span>
  <!---->
</div>"
`;
